### DISCUSSION TITLE
How to define type for useQuery

### DISCUSSION THREAD

**Discussion by dannywamuya** [trpc/trpc#4233] [Q&A]
So I'm building out my data table component and I would like to use trpc queries to fetch the data. I was previously using react-query while using REST in the table and the shape of the fetchFunction was this:
````ts
fetchFunction: (
    page?: number,
    size?: number
  ) => Promise<PaginatedResponse<T[]>>;
````

With that, I was able to fetch data like this: 
````ts
const { data, isFetching, refetch, isFetched } = useQuery<T[]>(
    [...dataKey, fetchDataOptions],
    async () => {
      const { data, totalPages } = await fetchFunction(pageIndex, pageSize);
      pages.current = totalPages;
      return data;
    },
    {
      initialData: [],
      keepPreviousData: true,
    }
  );
````

However, now that I am using trpc, I would like to just pass my query e.g `query={api.products.getAll.useQuery}` to my component and be able to handle the data in a modular manner inside my table. Knowing that I will always have the same return type when I call the query but still have my generic apply to the data returned of which data should be `T[]` i.e
````ts
const {data, isLoading} = query()
````
````
export interface DataTableProps<T> {
query: //type of query function
}
````

I would really appreciate any help in figuring this out. Thanks.

**Comment by Nick-Lucas** [comment]
Note that even if you're passing a hook in, you should keep the `useX` naming, because this is meaningful to eslint and react dev tools.

You have a couple options. For easy cases you can just take `ReturnType<typeof path.to.example.useQuery>`

For more complex cases you might want something like this: https://dev.to/nicklucas/trpc-patterns-router-factories-and-polymorphism-30b0#using-router-factories-polymorphically-in-react / https://trpc.io/docs/reactjs/infer-types#infer-abstract-types-from-a-router-factory
